acir/src/circuit/opcodes/black_box_function_call.rs

```diff
@@ -300,101 +438,33 @@ impl BlackBoxFuncCall {
         write_inputs(&self.get_inputs_vec(), &mut writer)?;
         write_outputs(&self.get_outputs_vec(), &mut writer)?;
 
-        Ok(())
+        match self {
+            BlackBoxFuncCall::Pedersen { domain_separator, .. } => {
+                write_u32(&mut writer, *domain_separator)?;
```

Comment:
The `domain_separator` coming after the outputs feels strange. Will it not be an input set by the user?

Reply:
If we move forward with custom methods for reading in the BB calls we should do the same for writing them. However, this will require us to match inside of the `write` function like we do for a `read`. This can be avoided though by using the strategy taken in this PR: https://github.com/noir-lang/acvm/pull/285/files

Reply:
It's a little different in the sense that it's a constant value, not a witness index, but yeah conceptually is also an input! Seeing that now some functions might have multiple variable length arrays of witness indices as inputs, I think I would go with fully custom (de)serialization methods per bb call yep

Reply:
We can avoid having custom write method using the strategy in PR #285 it seems though. Right now we are going to need a match statement in `write` like we have in `read`

Reply:
> constant value, not a witness index, but yeah conceptually is also an input

Like you said constants are still inputs and should encapsulated as part of the writing of the inputs imo

Reply:
But in that case we'd be serializing unnecesary length statements when the arguments are known-length, just to avoid, i feel that the option of specific write functions is more expressive. And if the match inside each write function is a problem (because it's a runtime check) we can type the content of each enum variant as a different struct, and provide a read/write trait impl for the structs that represent the data of each struct variant!

Reply:
I would want to get rid of the internal write method match statements. However, I don't think it is possible in Rust atm to impl on enum variants. We would have to impl on the enum itself. Which at that point requires us to match on the whole enum.

I'm going to try and see how I can avoid unnecessary length statement's by differentiating the function inputs with an enum